sched/config/config.go

```diff
@@ -3,445 +3,199 @@ package config
-type WorkerType PluginType
-type QueueType PluginType
-type PlannerType PluginType
-type RunnerType string //PluginType
-type ExecType string   //PluginType
-
-const (
-	//
-	ClusterTypeId   TypeId = "cluster"
-	SagaCoordTypeId TypeId = "sagaCoord"
-	WorkerTypeId    TypeId = "worker"
-	QueueTypeId     TypeId = "queue"
-	PlannerTypeId   TypeId = "planner"
-	RunnerTypeId    TypeId = "runner"
-	ExecTypeId      TypeId = "exec"
-	StatsTypeId     TypeId = "stats"
-	HealthTypeId    TypeId = "health"
-
-	//
-	ClusterMemory  ClusterType = "memory"
-	ClusterStatic  ClusterType = "static"
-	ClusterDynamic ClusterType = "dynamic"
-
-	//
-	QueueFake    QueueType = "fake"
-	QueueMemory  QueueType = "memory"
-	QueueDurable QueueType = "durable"
-
-	//
-	SagaCoordMemory  SagaCoordType = "memory"
-	SagaCoordDurable SagaCoordType = "durable"
-
-	//
-	WorkerMemory WorkerType = "memory"
-	WorkerThrift WorkerType = "thrift"
-
-	//
-	PlannerChaos  PlannerType = "chaos"
-	PlannerSimple PlannerType = "simple"
-	PlannerSmart  PlannerType = "smart"
-
-	//
-	RunnerSimple RunnerType = "simple"
-
-	//
-	ExecOs  ExecType = "os"
-	ExecSim ExecType = "sim"
-)
-
-func IntToPtr(i int) *int { return &i }
-
-/*
-
-
-## Example json. Any subset can be specified to override specific defaults.
-##
+func (p *Parser) Parse(configText []byte) (*Config, error) {
+	// TODO(dbentley): allow this to refer to a file instead of being a giant string
+	if len(configText) == 0 {
+		configText = emptyJson
+	}
+	var cfg topLevelConfig
+	err := json.Unmarshal(configText, &cfg)
+	if err != nil {
+		return nil, fmt.Errorf("Couldn't Parse top-level config: %v", err)
+	}
 
+	r := &Config{}
 
-{
-    "Cluster": {
-        "Type": "memory",
-        "Initial": 5,
-        "Count": 10
-    },
-    "Queue": {
-        "Type": "memory",
-        "Capacity": 1000
-    },
-    "SagaCoord": {
-        "Type": "memory"
-    },
-    "Worker": {
-        "Type": "memory",
-        "Runner": {
-            "Type": "simple",
-            "Exec": {
-                "Type": "os"
-            }
-        },
-        "PingIntervalMs": 0,
-        "TimeoutMs": 0
-    },
-    "Planner": {
-        "Type": "simple"
-    },
-    "Stats": {
-        "Path": "/admin/metrics.json",
-        "Enabled": true
-    },
-    "Health": {
-        "Path": "/",
-        "Ok": "OK",
-        "Enabled": true
-    }
-}
+	// Now we parse each config. To do this, we:
+	// 1) Parse its type
+	// 2) Find the FooConfig for this type
+	// 3) Unmarshal into the FooConfig
+	// 4) Set this config in the result
 
-## For posterity, original pseudo json.
-##
-{
- "cluster": {"type": "memory",  "initial": 5, "count": 10} ||
-            {"type": "static",  "initial": 5, "hosts": [":2345", ":2346"]} ||
-            {"type": "dynamic", "initial": 5, "max": 6},
-
- "queue": {"type": "fake", "count": 1000} ||
-          {"type": "memory", "capacity": 1000} ||
-          {"type": "durable"}
+	queueType, queueData := parseType(cfg.Queue)
```

Comment:
I hoisted out parseDomain, but it ends up being a little messy still because of Go's lack of generics; WDYT?